consul/state/query.go

```diff
@@ -0,0 +1,241 @@
+package state
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/consul/consul/structs"
+	"github.com/hashicorp/go-memdb"
+)
+
+// Queries is used to pull all the prepared queries from the snapshot.
+func (s *StateSnapshot) Queries() (memdb.ResultIterator, error) {
+	iter, err := s.tx.Get("queries", "id")
+	if err != nil {
+		return nil, err
+	}
+	return iter, nil
+}
+
+// Query is used when restoring from a snapshot. For general inserts, use
+// QuerySet.
+func (s *StateRestore) Query(query *structs.PreparedQuery) error {
+	if err := s.tx.Insert("queries", query); err != nil {
+		return fmt.Errorf("failed restoring query: %s", err)
+	}
+
+	if err := indexUpdateMaxTxn(s.tx, query.ModifyIndex, "queries"); err != nil {
+		return fmt.Errorf("failed updating index: %s", err)
+	}
+
+	s.watches.Arm("queries")
+	return nil
+}
+
+// QuerySet is used to create or update a prepared query.
+func (s *StateStore) QuerySet(idx uint64, query *structs.PreparedQuery) error {
+	tx := s.db.Txn(true)
+	defer tx.Abort()
+
+	// Call set on the Query.
+	if err := s.querySetTxn(tx, idx, query); err != nil {
+		return err
+	}
+
+	tx.Commit()
+	return nil
+}
+
+// querySetTxn is the inner method used to insert a prepared query with the
+// proper indexes into the state store.
+func (s *StateStore) querySetTxn(tx *memdb.Txn, idx uint64, query *structs.PreparedQuery) error {
+	// Check that the ID is set.
+	if query.ID == "" {
+		return ErrMissingQueryID
+	}
+
+	// Check for an existing query.
+	existing, err := tx.First("queries", "id", query.ID)
+	if err != nil {
+		return fmt.Errorf("failed query lookup: %s", err)
+	}
+
+	// Set the indexes.
+	if existing != nil {
+		query.CreateIndex = existing.(*structs.PreparedQuery).CreateIndex
+		query.ModifyIndex = idx
+	} else {
+		query.CreateIndex = idx
+		query.ModifyIndex = idx
+	}
+
+	// Verify that the name doesn't alias any existing ID. If we didn't do
```

Comment:
Beef up the comment with a better explanation about why this is needed / what's going on here.